### DISCUSSION TITLE
Expose app components & composables for extension use

### DISCUSSION THREAD

**Discussion by JoshTheDerf** [directus/directus#17234] [Feature Requests]
As it stands, many components under `app/src/views/private/components` and composables under `app/src/composables` are not available to app extensions.  

A number of these don't make sense for extension authors to use. However, some of them are critical for certain use cases.

For example, if someone wants to create a custom relational interface, they'll need the `<drawer-collection>` and `<drawer-item>` components for selecting and editing related rows. Without these components, selection and editing of relational data would have to be reimplemented from scratch for the custom interface. 

For some reason, the `<drawer-item>` component is registered in the Vue app and available to extensions, while `<drawer-collection>` is not. This can be done with a single change in `app/src/components/register.ts`, I'll submit a PR for this shortly.

**EDIT**: PR #17235

Additionally, exposing `app/composables/use-relation-*` would make implementation of custom relational interfaces much simpler and reduce extension boilerplate. However, the preferred way of exposing these is not immediately clear, nor is it clear how stable these composables are compared to the core stores backing them. 

Thoughts?

**Comment by rijkvanzanten** [comment]
The tricky bit with this is that that also means we have to be very careful with making any changes to those components, as any usage change to those components will now mean that extensions relying on it will have to be updated to match (eg it becomes a breaking change).

**Reply by JoshTheDerf** [reply]
This is true. At the same time, we're already exposing a number of such components implicitly with no indication of stability one way or another.

It might be worthwhile adding a warning in the documentation that anything not explicitly documented is an unstable API interface and may break unexpectedly, while still allowing extensions to be built using those components.

That provides a path forward: Once private parts of the code are documented/explicitly marked as stable, they can be "safely" relied upon. Up to that point, they're "use at your own risk". 

By the way, should I create an issue for the linked PR? I think that one is pretty safe given that it's just registering a component that's equivalent to ones already implicitly exposed to extensions.

**Comment by ghost** [comment]
I wanted to start developing a custom relational interface — a better/custom Builder (M2A) interface.

From what I've read here, it doesn't seem like it's really possible. @JoshTheDerf right?

@rijkvanzanten Why can’t developers easily extend directus and access all code available? As extension developers, we are responsible for making the extension work after an update of directus.

I’ve copied the [m2a interface](https://github.com/directus/directus/tree/main/app/src/interfaces/list-m2a) in order to create a custom extension based on it, but I don’t have access to every dependency/composable/… and it seems like it is a very long journey to rebuild this …

**Reply by JoshTheDerf** [reply]
So far copying and tweaking to `useStores` and `useAPI` in pretty much everything is all that's working for me. A few pitfalls due to `use*` using Vue `inject()`, but fairly straightforward apart from that. I haven't tackled `drawer-collection`, but I have rebuilt/modified `drawer-item` for our uses without _too_ much hassle.

Would definitely be much more preferable to have them exposed though.

**Comment by AxeemHaider** [comment]
@JonathanDoelan  Have you find out a way how we can use these **composable** and **view/private** in our extensions ?

**Reply by ghost** [reply]
I did not find a clean way to do it yet. Seems like directus wants us to go the dirty path copying single files from the repository and fixing all their dependencies manually …

**Reply by AxeemHaider** [reply]
By fixing manually you mean copy paste all these dependencies and those files also include dependencies then you have to fix them also ?

**Reply by ghost** [reply]
I guess so …

**Reply by JoshTheDerf** [reply]
> By fixing manually you mean copy paste all these dependencies and those files also include dependencies then you have to fix them also ?

Yep. Mainly because the composables and such import stores directly, but extensions can only access them through `useStores()`, which only works in a Vue context.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
